romtools/composite_vector_space/__init__.py

```diff
@@ -0,0 +1,154 @@
+            local_vector_space_dimensions = local_vector_space.get_basis().shape#get_dimension()
+            n_vars += local_vector_space_dimensions[0]
+            dims[i] = local_vector_space_dimensions[1]
+            total_number_of_bases += local_vector_space_dimensions[2]
+
+        nx = int(dims[0])
+        self.__n_vars = int(n_vars)
+        self.__nx = int(dims[0])
+        self.__total_number_of_bases = int(total_number_of_bases)
+        self.__extent = np.array([self.__n_vars,self.__nx,self.__total_number_of_bases],dtype='int')
+        assert np.allclose( np.diff(dims),np.zeros(dims.size-1)) , " Error constructing composite vector space, not all spaces have the same spatial dimension "
+
+        if variable_ordering is None:
+            variable_ordering = np.arange(0,n_vars,dtype='int')
+        else:
+            ## Check that we have a valid variable ordering
+            variables_left = np.arange(0,n_vars,dtype='int')
+            assert(len(variable_ordering) == n_vars)
+            for variable in variable_ordering:
+                assert np.isin(variable,variables_left)
+                variables_left = np.delete(variables_left,np.where(variable==variables_left))
+
+        ## Construct a global shift vector
+        self.__shift_vector = list_of_vector_spaces[0].get_shift_vector()
+        for i in range(1,self.n_vector_spaces_):
+          local_shift_vector = list_of_vector_spaces[i].get_shift_vector()
+          self.__shift_vector = np.append(self.__shift_vector,local_shift_vector,axis=0)
+
+        # Re-order according to variable ordering
+        self.__shift_vector = self.__shift_vector[variable_ordering]
+
+        ## Compute basis
+        if compact_representation == False:
+            # If False, return a global array
+            self.__basis = np.zeros((n_vars,nx,total_number_of_bases))
+            start_var_index = 0
+            start_basis_index = 0
+            for i in range(0,self.n_vector_spaces_):
+                local_vector_space = self.list_of_vector_spaces_[i]
+                local_vector_space_dimensions = local_vector_space.get_dimension()
+                local_basis = local_vector_space.get_basis()
+                dim = local_basis.shape
+                self.__basis[start_var_index:start_var_index+dim[0],:,start_basis_index:start_basis_index+dim[2]] = local_basis
+                start_var_index += dim[0]
+                start_basis_index += dim[2]
+            self.__basis = self.__basis[variable_ordering]
+
+        else:
+            # If true, return a list of the bases.
+            # This is much more efficient in terms of memory 
+            self.__compact_basis = []
+            self.__compact_shift_vector = []
+
+            for i in range(0,self.n_vector_spaces_):
+                self.__compact_basis.append(self.list_of_vector_spaces_[i].get_basis())
+                self.__compact_shift_vector.append(self.list_of_vector_spaces_[i].get_shift_vector())
+
```

Comment:
this construtor is giant, can we please break this up into smaller functions so that it is more readable?